translated/tech/20190807-go-thoughts-about-cobra.md

```diff
@@ -0,0 +1,309 @@
+
+## 轻量
+
+这个库的代码主要包含一个文件夹，而且很好理解，它不会影响你程序的性能。接下来，我们做一个压力测试（benchmark）：
+
+```go
+package main
+
+import (
+    "github.com/spf13/cobra"
+    "math/rand"
+    "os"
+    "strconv"
+    "testing"
+)
+
+func BenchmarkCmd(b *testing.B) {
+    for i := 0; i < b.N; i++ {
+        root := &cobra.Command{
+            Run: func(cmd *cobra.Command, args []string) {
+                println(`main`)
+            },
+            Use:   `test`,
+            Short: "test",
+        }
+
+        max := 100
+        for c := 0; c < max; c++ {
+            cmd := &cobra.Command{
+                Run: func(cmd *cobra.Command, args []string) {
+                    _ = c
+                },
+                Use:   `test-`+strconv.Itoa(c),
+                Short: `test `+strconv.Itoa(c),
+            }
+            root.AddCommand(cmd)
+        }
+
+        r := rand.Intn(max)
+        os.Args = []string{"go", "test-"+strconv.Itoa(r)}
+        _ = root.Execute()
+    }
+}
+```
+
+Cobra 运行 50 条命令只有 49.0μs 负载：
+
+```sh
+name   time/op
+Cmd-8  49.0µs ± 1%
+
+name   alloc/op
+Cmd-8  78.3kB ± 0%
+
+name   allocs/op
+Cmd-8     646 ± 0%
+```
+
+由于 Cobra 被设计运行在 CLI 模式下, 性能并不重要, 但是可以看出这个库有多么轻量.
+
+## 可替代性
+
+即使 Cobra 倾向于成为 Go 社区的标准包 - 浏览[最近使用 Cobra 的项目](https://github.com/spf13/cobra)证实了这一点 - 了解 Go 生态系统中有关 CLI 接口的内容总是好的。
+
+让我们回顾两个可以替代 Cobra 的项目：
+
+* [cli](https://github.com/urfave/cli)，一个用于构建命令行应用程序的包。这个包和 Cobra 一样流行，与嵌套命令，bash 补全，hook（钩子），alias(别名)等非常相似。但是，与 Cobra 不同，这个包使用 Go 库中的原生`flag`包。
+
+urfave/cli 例子：
+
+```go
+package main
+
+import (
+    "log"
+    "os"
+    "github.com/urfave/cli"
+)
+
+func main() {
+    app := cli.NewApp()
+    app.Commands = []cli.Command{
+        {
+            Action:  func(c *cli.Context) error {
+                println("Hello world")
+                return nil
+            },
+            Name:   `greet`,
+            Usage:  "This command will print Hello World",
+        },
+    }
+
+    err := app.Run(os.Args)
+    if err != nil {
+        log.Fatal(err)
+    }
+}
+```
+
+* [subcommands](https://github.com/google/subcommands)：虽然托管在 Google 的 Github 帐户中，但该项目并非官方 Google 产品。该库也很简单
+
+google/subcommands 示例
+
+```go
+package main
+
+import (
+    "context"
+    "flag"
+    "github.com/google/subcommands"
+)
+
+type GreetCommand struct {}
+func (g *GreetCommand) Name() string     { return "greet" }
+func (g *GreetCommand) Synopsis() string { return "Greet the world." }
+func (g *GreetCommand) Usage() string { return `Print Hello World.` }
+func (g *GreetCommand) SetFlags(*flag.FlagSet) {}
+func (p *GreetCommand) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+    println(`Hello World`)
+
+    return subcommands.ExitSuccess
+}
+
+func main() {
+    subcommands.Register(&GreetCommand{}, "foo")
+
+    flag.Parse()
+    subcommands.Execute(context.Background())
+}
+```
+
+该库基于一个接口而不是一个结构，如我们之前看到的 Cobra 或 CLI，并使代码更加冗长。
```

Comment:
The library is based on an interface rather than a structure, like Cobra or CLI that we have seen previously, and makes the code a bit more verbose.

The library is based on an interface rather than a structure, and makes the code a bit more verbose.
这是个并列的、有因果关系的句子，
a bit more 多一点点
该库基于一个接口而不是一个结构体，（因此）使代码稍显冗长。